coderd/workspacebuilds_test.go

```diff
@@ -1151,3 +1152,118 @@ func TestMigrateLegacyToRichParameters(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
+
+		_, err := regularUserClient.CreateWorkspaceBuild(ctx, workspace.ID, codersdk.CreateWorkspaceBuildRequest{
+			TemplateVersionID: workspace.LatestBuild.TemplateVersionID,
+			Transition:        codersdk.WorkspaceTransitionStart,
+			LogLevel:          "debug",
+		})
+
+		// Regular user: expect an error
+		require.NotNil(t, err)
+		var sdkError *codersdk.Error
+		isSdkError := xerrors.As(err, &sdkError)
+		require.True(t, isSdkError)
+		require.Contains(t, sdkError.Message, "Workspace builds with a custom log level are restricted to template authors only.")
+	})
+	t.Run("AsTemplateAuthor", func(t *testing.T) {
+		t.Parallel()
+
+		client := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+		user := coderdtest.CreateFirstUser(t, client)
+		version := coderdtest.CreateTemplateVersion(t, client, user.OrganizationID, &echo.Responses{
+			Parse:         echo.ParseComplete,
+			ProvisionPlan: echo.ProvisionComplete,
+			ProvisionApply: []*proto.Provision_Response{{
+				Type: &proto.Provision_Response_Log{
+					Log: &proto.Log{
+						Level:  proto.LogLevel_DEBUG,
+						Output: "want-it",
+					},
+				},
+			}, {
+				Type: &proto.Provision_Response_Log{
+					Log: &proto.Log{
+						Level:  proto.LogLevel_TRACE,
+						Output: "dont-want-it",
+					},
+				},
+			}, {
+				Type: &proto.Provision_Response_Log{
+					Log: &proto.Log{
+						Level:  proto.LogLevel_DEBUG,
+						Output: "done",
+					},
+				},
+			}, {
+				Type: &proto.Provision_Response_Complete{
+					Complete: &proto.Provision_Complete{},
+				},
+			}},
+		})
+		template := coderdtest.CreateTemplate(t, client, user.OrganizationID, version.ID)
+		coderdtest.AwaitTemplateVersionJob(t, client, version.ID)
+
+		// Create workspace
+		workspace := coderdtest.CreateWorkspace(t, client, user.OrganizationID, template.ID)
+		coderdtest.AwaitWorkspaceBuildJob(t, client, workspace.LatestBuild.ID)
+
+		// Create workspace build
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
+
+		build, err := client.CreateWorkspaceBuild(ctx, workspace.ID, codersdk.CreateWorkspaceBuildRequest{
+			TemplateVersionID: workspace.LatestBuild.TemplateVersionID,
+			Transition:        codersdk.WorkspaceTransitionStart,
+			ProvisionerState:  []byte(" "),
+			LogLevel:          "debug",
+		})
+		require.Nil(t, err)
+
+		build = coderdtest.AwaitWorkspaceBuildJob(t, client, build.ID)
+
+		// Watch for incoming logs
+		logs, closer, err := client.WorkspaceBuildLogsAfter(ctx, build.ID, 0)
+		require.NoError(t, err)
+		defer closer.Close()
+
+		for {
+			log, ok := <-logs
+			if !ok {
+				break
+			}
+
+			if log.Output == "dont-want-it" {
+				require.Failf(t, "unexpected log message", "%s log message shouldn't be logged: %s", log.Level, log.Output)
+			}
+
+			if log.Output == "done" {
+				return
+			}
+		}
```

Comment:
review: although it's in a naked for, this shouldn't hang forever as `WorkspaceBuildLogsAfter` exits on `<-ctx.Done()`

Reply:
Good catch! Refactored.